### ISSUE TITLE
Snapshot recovery should validate checksum if available

### ISSUE THREAD

**Issue by timvisee** [qdrant/qdrant#3371]
**Is your feature request related to a problem? Please describe.**
In <https://github.com/qdrant/qdrant/pull/2840> we've added checksums for snapshot files. The implementation is somewhat limited however, and requires further integration to make full use of it.

Currently, a malformed snapshot file can be recovered even if the integrity check with an attached checksum would fail. This is problematic.

**Describe the solution you'd like**
When a snapshot is being recovered, it should verify the attached checksum if any is provided. If this fails, recovery should be rejected.

Qdrant has a few REST and gRPC endpoints to trigger snapshot recovery. There are two flavors, recovering from an URI and by uploading a snapshot file directly. Both should get support for checksum integrity checks. These endpoints are described here: https://qdrant.tech/documentation/concepts/snapshots/#recover-via-api

A user should be able to (optionally) provide a checksum along with the request, so that Qdrant can verify its integrity. If the integrity check fails, we should clean up and return an error. If no checksum is provided we skip the integrity check and continue recovering like we do now.

A recovery request with checksum might look something like this:

```
PUT /collections/{collection_name}/snapshots/recover
{
  "location": "http://qdrant-node-1:6333/collections/{collection_name}/snapshots/snapshot-2022-10-10.shapshot",
  "checksum": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
}
```

So, in short:
- in all REST and gRPC snapshot recovery endpoints, add optional `checksum` field
- if a checksum is provided, Qdrant must verify the snapshot file integrity
- if integrity check with checksum failed, delete snapshot file and return an error

**Additional context**
There's other work to be done to properly integrate checksums, but that will be handled in different issues/PRs.

**Comment by tiagodavi** [comment]
> I'll try to do this one.

Are you still working on it ? 

I would like to solve it if you had given up.

**Comment by tiagodavi** [comment]
@timvisee Is there a way to build some sample data to test those endpoints locally? 

I ran `cargo test`, but it runs 17 tests only and it seems all the remaining tests are either `.sh` or `python` files?

Some tests seem to run Docker...

My goal is to generate some sample data with `migrations` in some way to inject fake data into the database, then be able to access the endpoint of that API, find where are the tests for this endpoint and improve them, change the api to include the optional checksum and test it individually, update thee documentation and send the PR.. 

What do you think?

**Comment by timvisee** [comment]
Implemented in <https://github.com/qdrant/qdrant/pull/3381>.
